### PR TITLE
Network reconstruction

### PR DESCRIPTION
**Pull request by jackzhhuang**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- Put the handshaking logic into the network crate from the network-p2p crate.
- Put the drive and core crates in the network-rpc crate into the network-p2p crate for reusability.
- Update the version of libp2p from 0.50 to 0.51.

### CODE REVIEW

File: network/src/network_business_handle.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+use std::pin::Pin;
+
+use bcs_ext::BCSCodec;
+use network_type::PeerInfo;
+use network_p2p::protocol::BusinessLayerHandle;
+use starcoin_types::startup_info::{ChainInfo, ChainStatus};
+
+impl BusinessLayerHandle for PeerInfo {
+    fn handshake(&self, peer_info: &[u8]) -> Result<(), (&'static str, String)> {
+        let other_chain_info = ChainInfo::decode(peer_info).unwrap();
```
Comment (nkysg) [inline network/src/network_business_handle.rs:13]: 这些地方unwrap()失败是不是直接导致节点崩溃了？

File: network-p2p/src/behaviour.rs [hunk 2]
```diff
@@ -62,7 +62,7 @@ pub enum BehaviourOut {
         protocol: Cow<'static, str>,
         /// Object that permits sending notifications to the peer.
         notifications_sink: NotificationsSink,
-        info: Box<ChainInfo>,
+        generic_data: Vec<u8>,
```
Comment (nkysg) [inline network-p2p/src/behaviour.rs:65]: 改成这个以后协议层是兼容的吗?如果不兼容是不是所有低版本的node节点都连不上了？

Comment (jackzhhuang) [reply network-p2p/src/behaviour.rs:65]: 只是重构代码，应该是兼容的，这个要测试一下

File: network/src/network_business_handle.rs (hunk 1 above)
Comment (jackzhhuang) [reply network/src/network_business_handle.rs:13]: 这个我后面改一改

**Review by sanlee42 — APPROVED** [review]

File: network-p2p/derive/Cargo.toml [hunk 3]
```diff
@@ -11,9 +11,9 @@ proc-macro = true
 authors = { workspace = true }
 edition = { workspace = true }
 license = { workspace = true }
-name = "network-rpc-derive"
+name = "network-p2p-derive"
 publish = { workspace = true }
-version = "1.13.5"
+version = "1.13.4"
```
Comment (nkysg) [inline network-p2p/derive/Cargo.toml:16]: 这里版本号怎么和其他地方不一样

Comment (jackzhhuang) [reply network-p2p/derive/Cargo.toml]: 已修改成1.13.5

File: network/src/service.rs [hunk 4]
```diff
@@ -695,10 +713,11 @@ impl Inner {
...
                 //2. Sync status change.
                 // may be update by repeat message, but can not find a more good way.
-                self.network_service.update_chain_status(ChainStatus::new(
-                    msg.compact_block.header.clone(),
-                    msg.block_info.clone(),
-                ));
+                self.network_service.update_business_status(
+                    ChainStatus::new(msg.compact_block.header.clone(), msg.block_info.clone())
+                        .encode()
+                        .unwrap(),
```
Comment (nkysg) [inline network/src/service.rs:719]: 这里ChainStatus::new变量是不是encode一定会成功？是的话这个unwrap()改成expect

File: network/src/network_p2p_handle.rs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                    rpc_protocols: status.rpc_protocols.to_vec(),
+                })
+            }
+            Err(_error) => Err(rep::FAILED_TO_ENCODE),
+        }
+    }
+}
+
+impl BusinessLayerHandle for Networkp2pHandle {
+    fn handshake(
```
Comment (sanlee42) [inline network/src/network_p2p_handle.rs:102]: 建议此处只对 status 进行处理，根据其返回状态来处理 notifications_sink和 CustomMessageOutcome，这两个在network这层不关心，放到 network-p2p里面比较合理。

File: network/src/service.rs (hunk 4 above)
Comment (jackzhhuang) [reply network/src/service.rs:719]: 好的，我改一下

**Review by nkysg — APPROVED** [review]

**Pull request merged**